pkg/controllers/podgroup/pg_controller.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2019 The Volcano Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package podgroup
+
+import (
+	"github.com/golang/glog"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/informers"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	kbv1 "volcano.sh/volcano/pkg/apis/scheduling/v1alpha1"
+	kbver "volcano.sh/volcano/pkg/client/clientset/versioned"
+	kbinfoext "volcano.sh/volcano/pkg/client/informers/externalversions"
+	kbinfo "volcano.sh/volcano/pkg/client/informers/externalversions/scheduling/v1alpha1"
+	kblister "volcano.sh/volcano/pkg/client/listers/scheduling/v1alpha1"
+)
+
+// Controller the Podgroup Controller type
+type Controller struct {
+	kubeClients kubernetes.Interface
+	kbClients   kbver.Interface
+
+	podInformer     coreinformers.PodInformer
+	pgInformer      kbinfo.PodGroupInformer
+	sharedInformers informers.SharedInformerFactory
+
+	// A store of pods
+	podLister corelisters.PodLister
+	podSynced func() bool
+
+	// A store of podgroups
+	pgLister kblister.PodGroupLister
+	pgSynced func() bool
+
+	queue workqueue.RateLimitingInterface
+}
+
+// NewPodgroupController create new Podgroup Controller
+func NewPodgroupController(
+	kubeClient kubernetes.Interface,
+	kbClient kbver.Interface,
+	schedulerName string,
+) *Controller {
+	cc := &Controller{
+		kubeClients: kubeClient,
+		kbClients:   kbClient,
+
+		queue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	cc.sharedInformers = informers.NewSharedInformerFactory(cc.kubeClients, 0)
+	cc.podInformer = cc.sharedInformers.Core().V1().Pods()
+	cc.podLister = cc.podInformer.Lister()
+	cc.podSynced = cc.podInformer.Informer().HasSynced
+	cc.podInformer.Informer().AddEventHandler(
+		cache.FilteringResourceEventHandler{
+			FilterFunc: func(obj interface{}) bool {
+				switch obj.(type) {
+				case *v1.Pod:
+					pod := obj.(*v1.Pod)
+					if pod.Spec.SchedulerName == schedulerName && len(pod.OwnerReferences) != 0 &&
+						(pod.Annotations == nil || pod.Annotations[kbv1.GroupNameAnnotationKey] == "") {
+						return true
+					}
+					return false
+				default:
```

Comment:
~There is another case, when delete event miss, please refer to `DeletedFinalStateUnknown`~

nvm， it only handles pod add